mindgraph/graph.py

```diff
@@ -26,8 +30,22 @@ def append(self, newnode) -> "Node":
         self._threads.append(newnode)
         return newnode
 
-    def pop(self, index: int) -> "Node":
+    def _linear_search(self, name: str) -> int:
+        """ find a node index by name in self._threads """
+        for i, node in enumerate(self._threads):
+            if node.name == name:
+                return i
+        return -1
+
+    def pop(self, item = None) -> "Node":
```

Comment:
Missing type annotation for item.

Reply:
It would probably be better if I wrote seperate methods for popByName and popByIndex.

Reply:
You know there's more possibilities than `typing.Any`?
In this case `item: typing.Optional[typing.Union[int, str]] = None` (or `typing.Optional[int, str, None]`) does the job.
Spoiler alert: mypy will complain about body of function, but there's fix for that as well.

Reply:
Yeah, i know about `Optional` but I wasn't able to get to work correctly.
`Optional[Union[int, str]]`:
```
graph.py:38: error: Argument 1 to "pop" of "list" has incompatible type "Union[int, str, None]"; expected "int"
graph.py:41: error: Argument 1 to "_linear_search" of "Node" has incompatible type "Union[int, str, None]"; expected "str"
```
`Optional[int, str, None]` yields:
```graph.py:35: error: Optional[...] must have exactly one type argument
graph.py:38: error: Argument 1 to "pop" of "list" has incompatible type "Optional[Any]"; expected "int"
graph.py:41: error: Argument 1 to "_linear_search" of "Node" has incompatible type "Optional[Any]"; expected "str"
```

In general I have had lot of trouble using `Optional` and `Union` in Mypy and I try to use as light as touch as possible with annotations.

Reply:
Oh, are those errors what you were referring to as your spoiler alert? What is the solution?

Reply:
First of all, I made mistake and of course the second option meant to be `Union[int, str, None]` - this is equivalent to `Optional[Union[int, str]]`. Optional with more than one argument isn't even possible to define, AFAIR.

And the solution is to check type with `if isinstance(item, int)` instead of `if type(item) == int`. That way mypy will understand, that inside if condition only `int` is possible, and after both returns, `int` and `None` are eliminated, so item must be `str`.

Reply:
One more note. You already know this from failing pep8 check, but for mypy it's also important that comparing to None is done by `if item is None` instead of `if item == None`. Otherwise even after second return it will still "think" type of item is `None` or `str` (`Optional[str]`)

Reply:
Thanks so much, I must have missed all these details when I read the mypy documentation. I have always had trouble using `optional` and `union`.

Reply:
I've fixed the annotations, however at this point my PR is behind master quite a bit. I'll try to resolve the conflicts today.